backend/fms_core/template_importer/row_handlers/library_qc/library_qc.py

```diff
@@ -15,43 +15,69 @@ def __init__(self):
         if not source_sample_obj.is_library:
-            self.errors['sample'] = f'The sample at the specified location is not a library: ${source_sample_obj.name}'
+            self.errors['sample'] = f'The sample at {barcode}@{coordinates} is not a library ({source_sample_obj.name})'
 
         # volumes
-        if measures['initial_volume'] is None:
-            self.errors['initial_volume'] = 'Initial volume must be specified.'
-        else:
-            change_in_volume = abs(measures['initial_volume'] - source_sample_obj.volume)
-            if (change_in_volume) > 0.0:
-                self.warnings['initial volume'] = f"The current library volume {source_sample_obj.volume }uL differs from the initial volume {measures['initial_volume']}uL"
-            
-        if measures['measured_volume'] is None:
-            self.errors['measured_volume'] = 'Measured volume must be specified.'
-        if measures['measured_volume'] < 0:
-            self.errors['measured_volume'] = f'Measured volume must be a positive value.'
-        # If the measured volume > initial volume, output a warning in case this is a user error.
-        if measures['measured_volume'] > measures['initial_volume']:
-            self.warnings['measured_volume'] = f"Measured volume {measures['measured_volume']} is greater than initial volume {measures['initial_volume']}"
-
-        if process_measurement['volume_used'] is None:
-            self.errors['volume_used'] = 'Volume used must be specified.'
-        if process_measurement['volume_used'] < 0:
-            self.errors['volume_used'] = 'Volume used must be a positive value.'
-
-        delta_volume = measures['measured_volume'] - measures['initial_volume']
-        final_volume = source_sample_obj.volume + delta_volume - process_measurement['volume_used']
-
-        change_in_initial_volume = abs(measures['initial_volume'] - source_sample_obj.volume)
-        if (change_in_initial_volume) > 0.0:
-            self.warnings['initial volume'] = f"The current library volume ({source_sample_obj.volume }uL) differs from the initial volume ({measures['initial_volume']}uL) in the template. The library volume will be set to {final_volume}uL."
+        initial_volume = measures['initial_volume']
+        measured_volume = measures['measured_volume']
+        volume_used = process_measurement['volume_used']
+        sample_volume = source_sample_obj.volume
+        final_volume = None
+        
+        if initial_volume is None:
+            self.errors['initial_volume'] = 'Initial volume must be specified'
+        if measured_volume is None:
+            self.errors['measured_volume'] = 'Measured volume must be specified'
+        if volume_used is None:
+            self.errors['volume_used'] = 'Volume used must be specified'
+        if sample_volume is None:
+            self.errors['library volume'] = 'Library volume is missing'
+
+        if not None in (initial_volume, measured_volume, volume_used, sample_volume):
```

Comment:
This is an ugly test ... you should check if there is errors instead ... Its much easier to understand and you are less likely to let it proceed because you missed something.